pyiron/atomistics/master/elastic.py

```diff
@@ -0,0 +1,385 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+from __future__ import print_function
+
+import numpy as np
+import scipy.constants
+from pyiron.atomistics.master.parallel import AtomisticParallelMaster
+from pyiron_base import JobGenerator
+from sklearn.linear_model import LinearRegression
+from collections import defaultdict
+import warnings
+import itertools
+
+__author__ = "Joerg Neugebauer, Jan Janssen"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Jan Janssen"
+__email__ = "janssen@mpie.de"
+__status__ = "production"
+__date__ = "Oct 21, 2020"
+
+
+eV_div_A3_to_GPa = (
+    1e21 / scipy.constants.physical_constants["joule-electron volt relationship"][0]
+)
+
+def _fit_coeffs_with_stress(
+        strain,
+        stress,
+        rotations,
+        max_polynomial_order=None,
+        fit_first_order=False,
+    ):
+    higher_terms = _get_higher_order_terms(
+        strain,
+        max_polynomial_order=max_polynomial_order,
+        rotations=rotations,
+        derivative=True,
+    )
+    strain = np.einsum('nik,mkl,njl->nmij',
+                       rotations, strain, rotations).reshape(-1, 9)
+    strain = strain[:, [0, 4, 8, 5, 2, 1]]
+    strain[:,3:] *= 2
+    if higher_terms is not None:
+        strain = np.concatenate((strain, higher_terms), axis=-1)
+    stress = np.einsum('nik,mkl,njl->nmij',
+                       rotations, stress, rotations).reshape(-1, 9)
+    stress = stress[:, [0, 4, 8, 5, 2, 1]]
+    score = []
+    coeff = []
+    for ii in range(6):
+        strain_tmp = strain.copy()
+        strain_tmp[:, 6:] *= strain[:, ii, np.newaxis]
+        reg = LinearRegression(
+            fit_intercept=fit_first_order).fit(strain_tmp, stress[:,ii])
+        coeff.append(reg.coef_)
+        score.append(reg.score(strain_tmp, stress[:,ii]))
+    return np.array(coeff), np.array(score)
+
+def _fit_coeffs_with_energies(
+        strain,
+        energy,
+        volume,
+        rotations,
+        max_polynomial_order=None,
+        fit_first_order=False,
+    ):
+    higher_terms = _get_higher_order_terms(
+        strain,
+        max_polynomial_order=max_polynomial_order,
+        rotations=rotations,
+        derivative=False,
+    )
+    strain_voigt = np.einsum('nik,mkl,njl->nmij',
+                             rotations, strain, rotations).reshape(-1, 9)
+    strain_voigt = strain_voigt[:, [0, 4, 8, 5, 2, 1]]
+    strain_voigt[:,3:] *= 2
+    energy = np.tile(energy, len(rotations))
+    volume = np.tile(volume, len(rotations))
+    strain = 0.5*np.einsum('ni,nj->nij', strain_voigt, strain_voigt)
+    strain = np.triu(strain).reshape(-1, 36)
+    strain = strain[:,np.sum(strain, axis=0)!=0]
```

Comment:
Three consecutive strain lines also smells like it needs a well-named function to summarize the operation that is being made

Reply:
Here I remove the lower triangle, which is the same as the upper triangle to stabilise the linear regression (what I do below with `coeff` is the opposite transformation). I add a few comments for now because I don't use the same functional elsewhere and I can't come up with a fancy function name makes it super clear.